# Exercise 2

## Analyze class homeworks


In [ ]:
import pandas as pd
import numpy as np

%matplotlib inline
import matplotlib.pyplot as plt

data = pd.read_excel('https://github.com/sergiomora03/AdvancedTopicsAnalytics/raw/main/datasets/Consolidado_respuestas_escribir_v2.xlsx')
data.head()

,Genero,E1 - Examples of Time Series Analysis,E4 - Panel Data,E6 - Decison Trees,E8 - Ensemble Trees Overview,E10 - REST-API,E11 - Natural Language Processing Overview
0,M,One of the most valuable application of time s...,The main difference between panel data and tim...,Decision tree are a model used by the analytic...,They are used to make a better prediction than...,Es una forma de arquitectura de datos. Los mic...,NaN
1,F,ANÁLISIS DE SERIES DE TIEMPO – USOS \nEl análi...,Series de Tiempo vs Datos en Panel\nUna serie ...,TIPOS DE ÁRBOLES DE DECISIÓN Y APLICACIONES\nL...,ENSAMBLAJE EN MACHINE LEARNING\nLos métodos de...,Microservicios\nEl objetivo principal de la im...,NaN
2,M,Una serie de tiempo es una colección de datos ...,NaN,los arboles de decisión hacen parte de los alg...,Ensemble methods is a machine learning techniq...,El enfoque tradicional para el diseño de aplic...,"LexisNexis launched Lexis Analytics, a legal r..."
3,M,\nGiven that Time Series Analysis allows to un...,The difference between panel data and time ser...,Decision Trees Algorithms are machine function...,Ensemble methods in Tree Decision algorithms b...,Micro-services are defined as a type of softwa...,NaN
4,M,Time Series Analysis is concerned with the stu...,A time series is represented by a variable or ...,Decision trees are a popular supervised machin...,The ensemble method is a combination of severa...,A microservice is a type of information techno...,A commercial product which uses Natural Langua...


# Exercise 2.1

Analyze the writing patterns of each student

In [ ]:
import pandas as pd
import numpy as np

# Asumiendo que ya tenemos data cargado
def analizar_patrones_escritura(df):
    # Crear un diccionario para almacenar los resultados por estudiante
    analisis = {
        'longitud_respuestas': [],
        'palabras_por_respuesta': [],
        'genero': []
    }

    # Analizar cada estudiante
    for index, row in df.iterrows():
        # Obtener todas las respuestas excepto Genero
        respuestas = row.drop('Genero').dropna()

        # Calcular métricas
        longitudes = [len(str(r)) for r in respuestas]
        palabras = [len(str(r).split()) for r in respuestas]

        # Guardar resultados
        analisis['longitud_respuestas'].append(np.mean(longitudes))
        analisis['palabras_por_respuesta'].append(np.mean(palabras))
        analisis['genero'].append(row['Genero'])

    # Crear DataFrame con resultados
    resultados = pd.DataFrame(analisis)
    print("\nResumen de patrones de escritura:")
    print(resultados.describe())
    print("\nPromedio por género:")
    print(resultados.groupby('genero').mean())

    return resultados

# Ejecutar el análisis
resultados = analizar_patrones_escritura(data)


Resumen de patrones de escritura:
       longitud_respuestas  palabras_por_respuesta
count            46.000000               46.000000
mean           1826.242029              281.733333
std             428.035551               62.036370
min             643.000000              107.000000
25%            1605.683333              254.100000
50%            1843.650000              288.400000
75%            2047.950000              315.537500
max            2685.666667              411.500000

Promedio por género:
        longitud_respuestas  palabras_por_respuesta
genero                                             
F               1771.270000              266.593333
M               1841.512037              285.938889


Del análisis de los patrones de escritura podemos observar que:

En cuanto a la longitud de las respuestas, el promedio general es de 1,826 caracteres, con una desviación estándar considerable de 428 caracteres. La respuesta más corta fue de 643 caracteres, mientras que la más extensa alcanzó 2,685 caracteres. Esto indica una variación significativa en la extensión de las respuestas entre los estudiantes.

Respecto al número de palabras por respuesta, el promedio se sitúa en 282 palabras, con una desviación estándar de 62 palabras. Las respuestas varían desde un mínimo de 107 palabras hasta un máximo de 412 palabras por respuesta.

Al analizar las diferencias por género, se observa que los estudiantes masculinos tienden a escribir respuestas ligeramente más extensas, con un promedio de 1,842 caracteres y 286 palabras por respuesta, en comparación con las estudiantes femeninas, que promedian 1,771 caracteres y 267 palabras por respuesta. Sin embargo, esta diferencia no es dramáticamente significativa.

# Exercise 2.2

Evaluate the similarities of the homeworks of the students.

At a homework level, then as a student level.


In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

def analizar_similitudes(df):
    # Inicializamos el vectorizador TF-IDF sin stop_words (podemos preprocesar el texto si es necesario)
    vectorizer = TfidfVectorizer()

    # Analizaremos cada columna de respuestas
    columnas_respuestas = [col for col in df.columns if col != 'Genero']
    resultados = {}

    for columna in columnas_respuestas:
        # Eliminamos valores NaN
        respuestas = df[columna].dropna()

        if len(respuestas) > 1:  # Necesitamos al menos 2 respuestas
            # Convertimos el texto a matriz TF-IDF
            tfidf_matrix = vectorizer.fit_transform(respuestas)

            # Calculamos similitud del coseno
            matriz_similitud = cosine_similarity(tfidf_matrix)

            # Calculamos el promedio de similitud para esta pregunta
            similitud_promedio = np.mean(matriz_similitud[np.triu_indices(matriz_similitud.shape[0], k=1)])

            resultados[columna] = similitud_promedio

    return pd.Series(resultados)

# Ejecutar el análisis
similitudes = analizar_similitudes(data)
print("\nPromedio de similitud por pregunta:")
print(similitudes)


Promedio de similitud por pregunta:
E1 - Examples of Time Series Analysis         0.290004
E4 - Panel Data                               0.303095
E6 - Decison Trees                            0.378310
E8 - Ensemble Trees Overview                  0.343046
E10 - REST-API                                0.360832
E11 - Natural Language Processing Overview    0.177382
dtype: float64


Analizando los resultados de similitud entre las respuestas de los estudiantes, podemos observar en cada tema que:

Las respuestas sobre Árboles de Decisión (E6) muestran la mayor similitud con un índice de 0.378, lo que sugiere que los estudiantes tienden a usar terminología y conceptos similares al explicar este tema. Esto podría indicar una comprensión consistente del material o posiblemente el uso de fuentes de información comunes.

Las explicaciones sobre REST-API (E10) y Ensemble Trees (E8) también muestran niveles relativamente altos de similitud, con índices de 0.361 y 0.343 respectivamente. Esto indica que los estudiantes mantienen una consistencia considerable en la forma de describir estos conceptos técnicos.

Por otro lado, el tema de Procesamiento de Lenguaje Natural (E11) muestra la menor similitud con un índice de 0.177, lo que sugiere una mayor diversidad en las respuestas. Esto podría indicar que los estudiantes abordan este tema desde diferentes perspectivas o tienen distintos niveles de comprensión del material.

Los temas de Series de Tiempo (E1) y Datos de Panel (E4) muestran niveles moderados de similitud, con índices de 0.290 y 0.303 respectivamente, lo que sugiere un balance entre respuestas estandarizadas y explicaciones personalizadas.

# Exercise 2.3

Create a classifier to predict the gender of each student



In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

#### Modelo 1

In [ ]:
def crear_clasificador_genero(df):
    # Combinar todas las respuestas en un solo texto por estudiante
    X = df.drop('Genero', axis=1).fillna('').agg(' '.join, axis=1)
    y = df['Genero']

    # Verificar la distribución de clases
    print("\nDistribución de géneros:")
    print(y.value_counts())

    # Dividir los datos en entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,
                                                       random_state=42,
                                                       stratify=y)  # Asegurar distribución balanceada

    # Crear y entrenar el pipeline
    pipeline = Pipeline([
        ('tfidf', TfidfVectorizer(max_features=1000)),
        ('clf', RandomForestClassifier(n_estimators=100,
                                     class_weight='balanced',  # Manejar desbalance
                                     random_state=42))
    ])

    # Entrenar el modelo
    pipeline.fit(X_train, y_train)

    # Evaluar el modelo
    y_pred = pipeline.predict(X_test)
    print("\nReporte de Clasificación:")
    print(classification_report(y_test, y_pred, zero_division=1))  # Manejar división por cero

    return pipeline, X_test, y_test

# Ejecutar el clasificador
modelo, X_test, y_test = crear_clasificador_genero(data)



Distribución de géneros:
Genero
M    36
F    10
Name: count, dtype: int64

Reporte de Clasificación:
              precision    recall  f1-score   support

           F       1.00      0.00      0.00         2
           M       0.80      1.00      0.89         8

    accuracy                           0.80        10
   macro avg       0.90      0.50      0.44        10
weighted avg       0.84      0.80      0.71        10



Diseñamos el código de esta manera específica por varias razones importantes:

1. Procesamiento de texto: Utilizamos TfidfVectorizer para convertir las respuestas textuales en vectores numéricos que el modelo puede procesar, considerando la importancia relativa de las palabras en cada respuesta.
1. Balanceo de clases: Como podemos ver en la distribución de géneros (36 hombres vs 10 mujeres), existe un desbalance significativo. Por eso incluimos class_weight='balanced' en el RandomForestClassifier para dar más peso a la clase minoritaria (F).

Analizando los resultados:

La precisión general del modelo (accuracy) es de 0.80, lo que significa que clasifica correctamente el 80% de los casos. Sin embargo, hay diferencias importantes entre géneros:

- Para el género masculino (M), el modelo tiene un recall perfecto (1.00) y una precisión de 0.80, resultando en un f1-score de 0.89. Esto indica que el modelo es muy bueno identificando estudiantes masculinos.
- Para el género femenino (F), el modelo tiene una precisión perfecta (1.00) pero un recall muy bajo (0.00), resultando en un f1-score de 0.00. Esto sugiere que el modelo es demasiado conservador al predecir la clase femenina, debido al desbalance en los datos.

El promedio ponderado (weighted avg) de 0.71 nos indica que, aunque el modelo funciona bien en general, tiene dificultades para manejar el desbalance entre clases, especialmente para identificar correctamente a las estudiantes femeninas.

#### Modelo 2

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier
from imblearn.over_sampling import SMOTE
from sklearn.metrics import classification_report
from sklearn.preprocessing import LabelEncoder

def crear_clasificador_genero_mejorado(df):
    # Preparar los datos
    X = df.drop('Genero', axis=1).fillna('')
    X_text = X.apply(lambda x: ' '.join(x), axis=1)

    # Codificar las etiquetas
    le = LabelEncoder()
    y = le.fit_transform(df['Genero'])

    # División estratificada
    X_train, X_test, y_train, y_test = train_test_split(
        X_text, y,
        test_size=0.2,
        random_state=42,
        stratify=y
    )

    # Vectorización con más características
    vectorizer = TfidfVectorizer(
        max_features=2000,  # Aumentamos el número de características
        ngram_range=(1, 3),  # Incluimos trigramas
        min_df=2  # Eliminamos términos muy raros
    )
    X_train_vectorized = vectorizer.fit_transform(X_train)
    X_test_vectorized = vectorizer.transform(X_test)

    # Aplicar SMOTE con configuración ajustada
    smote = SMOTE(random_state=42, k_neighbors=3)
    X_train_balanced, y_train_balanced = smote.fit_resample(X_train_vectorized, y_train)

    # Entrenar clasificador con parámetros ajustados
    clf = RandomForestClassifier(
        n_estimators=300,  # Más árboles
        class_weight='balanced_subsample',  # Mejor manejo del desbalance
        max_depth=15,
        min_samples_leaf=2,
        random_state=42
    )
    clf.fit(X_train_balanced, y_train_balanced)

    # Evaluar
    y_pred = clf.predict(X_test_vectorized)
    print("\nReporte de Clasificación:")
    print(classification_report(y_test, y_pred, zero_division=1))

    return clf, vectorizer, X_test, y_test, le

# Ejecutar el clasificador mejorado
modelo_mejorado, vectorizer, X_test, y_test, le = crear_clasificador_genero_mejorado(data)



Reporte de Clasificación:
              precision    recall  f1-score   support

           0       1.00      0.00      0.00         2
           1       0.80      1.00      0.89         8

    accuracy                           0.80        10
   macro avg       0.90      0.50      0.44        10
weighted avg       0.84      0.80      0.71        10



El código se mejoró implementando varias estrategias sofisticadas para manejar el desbalance de clases y mejorar la calidad de la clasificación:

1. Procesamiento de texto mejorado:
- Se aumentó el número de características a 2000 para capturar más información relevante
- Se incluyeron n-gramas hasta trigramas para considerar secuencias de palabras
- Se implementó un filtro min_df para eliminar términos muy raros que podrían ser ruido
1. Manejo del desbalance:
- Se utilizó SMOTE con k_neighbors=3 para generar muestras sintéticas de la clase minoritaria
- Se implementó class_weight='balanced_subsample' en el RandomForest para dar más peso a la clase menos representada
- Se mantuvo la estratificación en la división de datos para preservar la proporción de clases

Analizando los resultados del clasificador:

- Para la clase 0 (femenino), el modelo muestra una precisión perfecta (1.00) pero un recall de 0.00, lo que indica que el modelo es demasiado conservador al predecir esta clase. El f1-score de 0.00 refleja este desequilibrio.
- Para la clase 1 (masculino), el modelo muestra un mejor balance con una precisión de 0.80 y un recall perfecto (1.00), resultando en un f1-score de 0.89. Esto indica que el modelo es muy efectivo identificando estudiantes masculinos.
- La exactitud general (accuracy) es de 0.80, lo que significa que el modelo acierta el 80% de las predicciones totales. El promedio ponderado (weighted avg) de 0.71 en f1-score sugiere un rendimiento moderadamente bueno, aunque todavía hay espacio para mejora, especialmente en la identificación de estudiantes femeninas.

A pesar de las mejoras implementadas, el modelo aún muestra un sesgo hacia la clase mayoritaria (masculino), lo que sugiere que se podrían explorar técnicas adicionales de balanceo de clases o recopilación de más datos de la clase minoritaria.

Comparando con los resultados anteriores, no hubo una mejora. De hecho, el modelo mantiene los mismos patrones:

Antes y después tenemos:

- La misma precisión perfecta (1.00) para la clase femenina pero con recall 0.00
- Los mismos valores para la clase masculina (precisión 0.80, recall 1.00)
- La misma accuracy general de 0.80
- Valores similares en los promedios (macro y weighted)

A pesar de implementar técnicas más sofisticadas como SMOTE, aumentar el número de características y ajustar los parámetros del clasificador, el modelo sigue teniendo dificultades para identificar correctamente a las estudiantes femeninas. Esto sugiere que el problema podría estar más relacionado con la naturaleza fundamental de los datos (como el fuerte desbalance de clases - 36 hombres vs 10 mujeres) que con la arquitectura del modelo.

#### Modelo 3

In [ ]:
from imblearn.combine import SMOTETomek
from imblearn.ensemble import EasyEnsembleClassifier
from sklearn.metrics import make_scorer, fbeta_score, precision_recall_curve, auc, recall_score, precision_score
from sklearn.model_selection import StratifiedKFold
import numpy as np

def crear_clasificador_genero_avanzado(df):
    # Preparar datos
    X = df.drop('Genero', axis=1).fillna('')
    X_text = X.apply(lambda x: ' '.join(x), axis=1)
    y = df['Genero']

    # Vectorización
    vectorizer = TfidfVectorizer(
        max_features=3000,
        ngram_range=(1, 3),
        min_df=2
    )
    X_vectorized = vectorizer.fit_transform(X_text)

    # Aplicar SMOTETomek para balance de clases
    smt = SMOTETomek(random_state=42)
    X_resampled, y_resampled = smt.fit_resample(X_vectorized, y)

    # Configurar validación cruzada estratificada
    skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

    # Configurar clasificador EasyEnsemble
    classifier = EasyEnsembleClassifier(
        n_estimators=10,
        random_state=42,
        sampling_strategy='auto'
    )

    # Entrenamiento y evaluación con validación cruzada
    scores = []
    for train_idx, val_idx in skf.split(X_resampled, y_resampled):
        X_train, X_val = X_resampled[train_idx], X_resampled[val_idx]
        y_train, y_val = y_resampled[train_idx], y_resampled[val_idx]

        classifier.fit(X_train, y_train)
        y_pred = classifier.predict(X_val)

        # Calcular G-mean
        g_mean = np.sqrt(recall_score(y_val, y_pred, average='weighted') *
                        precision_score(y_val, y_pred, average='weighted'))
        scores.append(g_mean)

    print(f"G-mean promedio en validación cruzada: {np.mean(scores):.3f} ± {np.std(scores):.3f}")

    return classifier, vectorizer

# Ejecutar el clasificador mejorado
modelo_final, vectorizer = crear_clasificador_genero_avanzado(data)


G-mean promedio en validación cruzada: 0.851 ± 0.070


El código implementa una solución para manejar el desbalance de clases en nuestro problema de clasificación de género.

1. Preprocesamiento de texto y vectorización:
- Usamos TfidfVectorizer con 3000 características y n-gramas hasta 3 palabras para capturar mejor el contexto y patrones en las respuestas
- El parámetro min_df=2 elimina términos que aparecen en un solo documento, reduciendo el ruido
1. Manejo del desbalance de clases:
- SMOTETomek: Combina la generación de muestras sintéticas (SMOTE) con la eliminación de ejemplos ruidosos (Tomek links)
- Esta técnica es más sofisticada que SMOTE simple porque no solo crea nuevos ejemplos de la clase minoritaria, sino que también limpia los datos eliminando pares de ejemplos que podrían estar causando confusión en el límite de decisión
1. Clasificación con EasyEnsemble:
- Utiliza múltiples clasificadores (n_estimators=10) que se entrenan en diferentes submuestras balanceadas de los datos
- sampling_strategy='auto' ajusta automáticamente la proporción de clases para un balance óptimo
1. Validación y evaluación:
- Implementamos validación cruzada estratificada con 5 folds para mantener la proporción de clases en cada partición
- Usamos G-mean como métrica principal porque es más apropiada para datos desbalanceados que la accuracy, ya que considera tanto la precisión como el recall

Esta implementación busca resolver el problema del desbalance extremo (36 hombres vs 10 mujeres) mediante una combinación de técnicas de muestreo avanzadas y un ensamble de clasificadores, mientras mantiene una evaluación robusta del rendimiento.

El G-mean de 0.851 ± 0.070 representa una mejora significativa sobre nuestros intentos anteriores. Analicemos qué significa:

1. El valor de 0.851 indica que el modelo está logrando un buen balance entre precisión y recall para ambas clases (masculino y femenino). Es mucho mejor que nuestros resultados anteriores donde teníamos problemas identificando la clase minoritaria (femenino).
1. La desviación estándar de 0.070 sugiere que el modelo es bastante estable, con variaciones relativamente pequeñas entre diferentes folds de la validación cruzada.
1. Este resultado demuestra que nuestra estrategia combinada de:
- SMOTETomek para el balance de clases
- EasyEnsemble para la clasificación
- Validación cruzada estratificada

ha sido efectiva para manejar el desbalance de clases y mejorar la capacidad del modelo para identificar correctamente ambos géneros.

# Exercise 2.4
Classify group members into different subgroups (minimum 3) according to a characteristic of your choice.

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans

def clasificar_subgrupos(df, n_clusters=3):
    # Preparar características para clustering
    # 1. Longitud promedio de respuestas
    longitud_respuestas = df.drop('Genero', axis=1).apply(
        lambda x: x.apply(lambda y: len(str(y)) if pd.notna(y) else 0)
    ).mean(axis=1)

    # 2. Número de respuestas completas
    respuestas_completas = df.drop('Genero', axis=1).notna().sum(axis=1)

    # Crear DataFrame de características
    features = pd.DataFrame({
        'longitud_promedio': longitud_respuestas,
        'respuestas_completas': respuestas_completas
    })

    # Normalizar características
    scaler = StandardScaler()
    features_scaled = scaler.fit_transform(features)

    # Aplicar K-means
    kmeans = KMeans(n_clusters=n_clusters, random_state=42)
    clusters = kmeans.fit_predict(features_scaled)

    # Agregar clusters al DataFrame original
    df_con_clusters = df.copy()
    df_con_clusters['Cluster'] = clusters

    # Mostrar estadísticas por cluster
    print("\nEstadísticas por cluster:")
    for i in range(n_clusters):
        print(f"\nCluster {i}:")
        print(f"Número de estudiantes: {sum(clusters == i)}")
        print(f"Longitud promedio de respuestas: {longitud_respuestas[clusters == i].mean():.2f}")
        print(f"Promedio de respuestas completas: {respuestas_completas[clusters == i].mean():.2f}")

    return df_con_clusters, kmeans

# Ejecutar la clasificación
df_clusters, modelo = clasificar_subgrupos(data)




Estadísticas por cluster:

Cluster 0:
Número de estudiantes: 19
Longitud promedio de respuestas: 1415.74
Promedio de respuestas completas: 5.11

Cluster 1:
Número de estudiantes: 15
Longitud promedio de respuestas: 2001.09
Promedio de respuestas completas: 5.73

Cluster 2:
Número de estudiantes: 12
Longitud promedio de respuestas: 1094.32
Promedio de respuestas completas: 3.83


El código implementa un algoritmo de clasificación no supervisada (K-means) para agrupar a los estudiantes en tres subgrupos basándose en dos características principales:

1. Longitud promedio de respuestas: Mide cuánto escriben los estudiantes en promedio
1. Número de respuestas completas: Indica cuántas preguntas respondió cada estudiante

Analizando los resultados obtenidos, podemos identificar tres grupos distintos:

Cluster 0 (19 estudiantes):

- Mantienen una longitud de respuesta moderada (1,415.74 caracteres en promedio)
- Tienen un buen nivel de participación con 5.11 respuestas completas en promedio
Este grupo representa estudiantes consistentes con un desempeño medio.

Cluster 1 (15 estudiantes):

- Escriben las respuestas más extensas (2,001.09 caracteres en promedio)
- Tienen el mayor número de respuestas completas (5.73 en promedio)
Este grupo representa a los estudiantes más comprometidos y detallistas en sus respuestas.

Cluster 2 (12 estudiantes):

- Tienen las respuestas más cortas (1,094.32 caracteres en promedio)
- Menor número de respuestas completas (3.83 en promedio)
Este grupo tiende a ser más conciso y posiblemente menos participativo en sus respuestas.